### ISSUE TITLE
where exactly teacher-forcing implemented in decoder?

### ISSUE THREAD

**Issue by Skumarr53** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#104]
Thank you for your implementation of this paper. I am new in this field, In fact, this is my first PyTorch project. I got the most of it. However, I couldn't able to figure out where exactly the **teacher-forcing** is implemented in the decoder part. And I know that **teacher-forcing** not implemented in eval.py but don't know where.

Please could you point out to the line in both decoder.py and eval.py where **teacher-forcing** is implemented and not implemented respectively?

**Comment by kmario23** [comment]
Hi,
  Teacher forcing means we ignore/discard the predictions of the model at each time step during training. So, when we give a batch size of tokens as input to the model, it will generate a batch size of tokens as output (it's only batch size since we don't beam search while training the model parameters). However, we will not care about these output tokens because we have the actual tokens already. So, we just use that.

The line you're after for teacher forcing is when you obtain the input tensor at each time step (i.e. `t`) and it is at:  [`models.py#L208`](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/master/models.py#L208). Observe how the input tensor is always taken from the embeddings which in turn is obtained using the original indices that we pass (i.e. `encoded_captions`).

In the *real* evaluation mode, we usually will not have a ground truth caption. (Although here, and in all supervised learning tasks, we do have ground truth captions for the val/test set). So, there is no way to teacher force the model during the testing mode, actually it doesn't make any sense to do so, since we want to caption images out-in-the-wild. So, we use beam search to obtain the next `k` tokens (i.e. their indices) at each time step `t` . For this, look at the variable `next_word_inds` in `eval.py` and how it is set to `k_prev_words` at each time step at the end of the `while` loop. Subsequently, these `k_prev_words` will be used to get the embeddings ([`eval.py#L101`](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/master/eval.py#L101)) which will then be passed to the decoder ([`eval.py#L108`](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/master/eval.py#L108)). This ride continues until we generate an `<end>` token, at which point we're done with generating a caption for the current image and proceed to the next one, until we exhaust all the images in our test set.

**Comment by Skumarr53** [comment]
@kmario23 Thanks for your response. I overlooked the `embedding` part, thanks for pointing it out.
